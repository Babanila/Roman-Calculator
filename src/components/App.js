import React, { Component, useState, useEffect } from "react";
import useDeepCompareEffect from "use-deep-compare-effect";
import styled from "@emotion/styled";
import KeypadButton from "./KeypadButton";
import Display from "./Display";
import { convertToRoman, convertToNumber } from "./Helpers";

const AppDiv = styled.div`
    margin: 0;
    padding: 0;
    width: 100%;
    max-height: 100%;
    box-sizing: border-box;
    font-family: Gotham Rounded, sans-serif;
    font-weight: normal;
    display: grid;
    justify-content: center;
    align-content: center;
`;

const HeadingDiv = styled.div`
    width: 100%;
    display: flex;
    flex-direction: row;
    text-align: center;
    justify-content: center;
    margin: 0 auto;
    z-index: 100;
    @media (min-width: 0px) and (max-width: 400px) {
        font-size: 1rem;
        height: 3rem;
        padding: 0;
    }
`;
// display: none;

const CalculatorDiv = styled.div`
    max-height: 100%;
    border: 1px solid #080808;
`;

function App() {
    const [prevOperation, setPrevOperation] = useState(" ");
    const [currentOperation, setCurrentOperation] = useState(" ");
    const [operator, setOperator] = useState(" ");
    const baseArray = {
        M: 1000,
        CM: 900,
        D: 500,
        CD: 400,
        C: 100,
        XC: 90,
        L: 50,
        XL: 40,
        X: 10,
        IX: 9,
        VIII: 8,
        VII: 7,
        VI: 6,
        V: 5,
        IV: 4,
        I: 1,
    };
    // Reset to default
    const handleReset = () => {
        setPrevOperation(" ");
        setCurrentOperation(" ");
        setOperator(" ");
    };

    // Delete or remove the last input
    const handleDelete = message => {
        const newMessage = message.toString().slice(0, -1);
        setCurrentOperation(newMessage);
    };

    const handleFirstOperand = operatorInput => {
        if (prevOperation === " ") {
            setOperator(operatorInput);
            setPrevOperation(currentOperation);
            setCurrentOperation("");
        } else {
            handleCompute(prevOperation, currentOperation, operatorInput, baseArray);
        }
    };

    // Compute handler for arithmetics operations
    const handleOperation = (prev, current, opera) => {
        let computeResult;
        switch (opera) {
            case "+":
                computeResult = prev + current;
                break;
            case "-":
                computeResult = prev - current;
                break;
            case "*":
                computeResult = prev * current;
                break;
            case "รท":
                computeResult = prev / current;
                break;
            default:
                computeResult = 0;
        }
        return computeResult;
    };

    const handleCompute = (first, second, joiner, arr) => {
        const num1 = convertToNumber(arr, first);
        const num2 = convertToNumber(arr, second);
        const combineNum = handleOperation(num1, num2, joiner);
        if (combineNum <= 0) return "NaN";
        const result = convertToRoman(arr, combineNum);
        return result;
    };

    const handleClick = e => {
        e.preventDefault();
        const { value } = e.target;
        const stringifyValue = value.toString();

        if (stringifyValue === "AC") return handleReset();

        if (stringifyValue === "DEL") return handleDelete(currentOperation);

        // Disable multiple entering of operator(operand) when second parameters is not provided
        if (
            (stringifyValue === "+" ||
                stringifyValue === "-" ||
                stringifyValue === "*" ||
                stringifyValue === "รท") &&
            (currentOperation.includes("+", "-", "*", "รท", "") || currentOperation === " ")
        )
            return;

        // Creating first parameter
        if (
            (stringifyValue === "+" ||
                stringifyValue === "-" ||
                stringifyValue === "*" ||
                stringifyValue === "รท") &&
            currentOperation !== " "
        ) {
            return handleFirstOperand(stringifyValue);
        }

        // Handle input from user
        currentOperation == " "
            ? setCurrentOperation(stringifyValue)
            : setCurrentOperation(currentOperation + stringifyValue);

        if (stringifyValue === "=") {
            setCurrentOperation(
                handleCompute(prevOperation, currentOperation, operator, baseArray)
            );
            setPrevOperation(" ");
            setOperator(" ");
        }
    };

    return (
        <AppDiv>
            <HeadingDiv>
                <h2>Roman Calculator</h2>
            </HeadingDiv>
            <CalculatorDiv>
                <Display
                    prevOperation={prevOperation}
                    currentOperation={currentOperation}
                    operator={operator}
                />
                <KeypadButton onClick={handleClick} />
            </CalculatorDiv>
        </AppDiv>
    );
}

export default App;
