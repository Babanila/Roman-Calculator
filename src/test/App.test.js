import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { render, cleanup } from "@testing-library/react";
import { shallow, mount } from "enzyme";
import App from "../components/App";

describe("App Component Test", () => {
    it("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(<App />, div);
        ReactDOM.unmountComponentAtNode(div);
    });

    it("should renders without crashing", () => {
        const AppComponent = renderer.create(<App />).toJSON();
        expect(AppComponent).toMatchSnapshot();
    });

    it("should render element correctly", () => {
        const AppWrapper = shallow(<App />);

        expect(AppWrapper.find("div").exists).toBeTruthy();
        expect(AppWrapper.find("h2").exists).toBeTruthy();
        expect(AppWrapper.find("Display").exists).toBeTruthy();
        expect(AppWrapper.find("KeypadButton").exists).toBeTruthy();
        expect(AppWrapper.find("div").length).toEqual(0);
        expect(AppWrapper.find("h2").length).toEqual(1);
        expect(AppWrapper.find("KeypadButton").length).toEqual(1);
        expect(AppWrapper.find("Display").length).toEqual(1);
    });

    it("should render correct number of element", () => {
        const AppWrapper = mount(<App />);

        expect(AppWrapper.find("div").exists).toBeTruthy();
        expect(AppWrapper.find("h2").exists).toBeTruthy();
        expect(AppWrapper.find("Display").exists).toBeTruthy();
        expect(AppWrapper.find("KeypadButton").exists).toBeTruthy();
        expect(AppWrapper.find("div").length).toEqual(7);
        expect(AppWrapper.find("h2").length).toEqual(1);
        expect(AppWrapper.find("Display").length).toEqual(1);
        expect(AppWrapper.find("KeypadButton").length).toEqual(1);
    });
});
