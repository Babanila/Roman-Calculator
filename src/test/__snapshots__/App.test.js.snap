// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component Test should renders without crashing 1`] = `
<div
  className="css-198qf24"
>
  <div
    className="css-ub5z1o"
  >
    <h2>
      Roman Calculator
    </h2>
  </div>
  <div
    className="css-jsz1sl"
  >
    <div
      className="css-vb37s6"
    >
      <div
        className="css-nrsn95"
        data-previous-operand={true}
      >
         
         
         
      </div>
      <div
        className="css-1jxc83g"
        data-current-operand={true}
      >
         
      </div>
    </div>
    <div
      className="css-1u2c2og"
    >
      <button
        className="basic-Button css-1kddh5u"
        data-label="I"
        onClick={[Function]}
        value="I"
      >
        I
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="II"
        onClick={[Function]}
        value="II"
      >
        II
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="III"
        onClick={[Function]}
        value="III"
      >
        III
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="+"
        onClick={[Function]}
        value="+"
      >
        +
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="IV"
        onClick={[Function]}
        value="IV"
      >
        IV
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="V"
        onClick={[Function]}
        value="V"
      >
        V
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="VI"
        onClick={[Function]}
        value="VI"
      >
        VI
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="-"
        onClick={[Function]}
        value="-"
      >
        -
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="VII"
        onClick={[Function]}
        value="VII"
      >
        VII
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="VIII"
        onClick={[Function]}
        value="VIII"
      >
        VIII
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="IX"
        onClick={[Function]}
        value="IX"
      >
        IX
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="รท"
        onClick={[Function]}
        value="รท"
      >
        รท
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="X"
        onClick={[Function]}
        value="X"
      >
        X
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="L"
        onClick={[Function]}
        value="L"
      >
        L
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="C"
        onClick={[Function]}
        value="C"
      >
        C
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="*"
        onClick={[Function]}
        value="*"
      >
        *
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="D"
        onClick={[Function]}
        value="D"
      >
        D
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="M"
        onClick={[Function]}
        value="M"
      >
        M
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="AC"
        onClick={[Function]}
        value="AC"
      >
        AC
      </button>
      <button
        className="basic-Button css-1kddh5u"
        data-label="="
        onClick={[Function]}
        value="="
      >
        =
      </button>
      <button
        className="basic-Button css-adl3hz"
        data-label="DELETE"
        onClick={[Function]}
        value="DELETE"
      >
        DELETE
      </button>
    </div>
  </div>
</div>
`;
